File: matlab/experiments/plot_results.m
function plot_results()
% PLOT_RESULTS Generate publication-quality figures from experiment results.
%
%   Loads results from data/results/ and creates figures suitable for
%   IEEE conference/journal submission.

    fprintf('=== Generating Publication Figures ===\n');

    results_dir = '../../data/results';

    % Set default figure properties for publication
    set(0, 'DefaultAxesFontSize', 12);
    set(0, 'DefaultTextFontSize', 12);
    set(0, 'DefaultLineLineWidth', 1.5);

    % --- Figure 1: Baseline Comparison Bar Chart ---
    if exist(fullfile(results_dir, 'baseline_comparison.mat'), 'file')
        load(fullfile(results_dir, 'baseline_comparison.mat'), 'results', 'methods_list');

        rmse_values = [results.odometry.mean_rmse, ...
                       results.ekf_fixed_40.mean_rmse, ...
                       results.ekf_fixed_60.mean_rmse, ...
                       results.ekf_fixed_80.mean_rmse, ...
                       results.ekf_fixed_100.mean_rmse, ...
                       results.adaptive.mean_rmse];

        fig1 = figure('Name', 'Baseline Comparison', 'Position', [100 100 800 400]);
        b = bar(rmse_values);
        b.FaceColor = 'flat';
        b.CData(end, :) = [0.2 0.6 0.2];  % Green for proposed method
        set(gca, 'XTickLabel', methods_list, 'XTickLabelRotation', 30);
        ylabel('Mean RMSE (m)');
        title('Building Localization: Method Comparison');
        grid on;

        saveas(fig1, fullfile(results_dir, 'fig_baseline_comparison.png'));
        fprintf('  Saved: fig_baseline_comparison.png\n');
    end

    % --- Figure 2: Observability Validation ---
    if exist(fullfile(results_dir, 'observability_validation.mat'), 'file')
        load(fullfile(results_dir, 'observability_validation.mat'), 'results');
        ts = results.timestamps;

        fig2 = figure('Name', 'Observability Validation', 'Position', [100 100 800 600]);

        subplot(2,1,1);
        hold on;
        fill([ts, fliplr(ts)], ...
             [mean(results.circular.errors,1) + std(results.circular.errors,0,1), ...
              fliplr(mean(results.circular.errors,1) - std(results.circular.errors,0,1))], ...
             [0.7 0.7 1], 'EdgeColor', 'none', 'FaceAlpha', 0.5);
        fill([ts, fliplr(ts)], ...
             [mean(results.straight.errors,1) + std(results.straight.errors,0,1), ...
              fliplr(mean(results.straight.errors,1) - std(results.straight.errors,0,1))], ...
             [1 0.7 0.7], 'EdgeColor', 'none', 'FaceAlpha', 0.5);
        plot(ts, mean(results.circular.errors,1), 'b-', 'LineWidth', 2);
        plot(ts, mean(results.straight.errors,1), 'r-', 'LineWidth', 2);
        xlabel('Time (s)'); ylabel('Position Error (m)');
        title('EKF Convergence by Trajectory Type');
        legend('', '', 'Circular orbit', 'Straight line (degenerate)', 'Location', 'northeast');
        grid on; hold off;

        subplot(2,1,2);
        hold on;
        plot(ts, mean(results.circular.traces,1), 'b-', 'LineWidth', 2);
        plot(ts, mean(results.straight.traces,1), 'r-', 'LineWidth', 2);
        xlabel('Time (s)'); ylabel('trace(P)');
        title('Covariance Trace Over Time');
        legend('Circular orbit', 'Straight line', 'Location', 'northeast');
        set(gca, 'YScale', 'log');
        grid on; hold off;

        saveas(fig2, fullfile(results_dir, 'fig_observability_validation.png'));
        fprintf('  Saved: fig_observability_validation.png\n');
    end

    % --- Figure 3: Noise Sensitivity ---
    if exist(fullfile(results_dir, 'noise_sensitivity.mat'), 'file')
        load(fullfile(results_dir, 'noise_sensitivity.mat'), 'results');

        fig3 = figure('Name', 'Noise Sensitivity', 'Position', [100 100 600 400]);
        errorbar(results.sigma_values, results.mean_rmse, results.std_rmse, ...
                 'bo-', 'LineWidth', 2, 'MarkerSize', 8, 'MarkerFaceColor', 'b');
        xlabel('Pixel Noise \sigma (px)');
        ylabel('Final RMSE (m)');
        title('Localization Accuracy vs. Measurement Noise');
        grid on;

        saveas(fig3, fullfile(results_dir, 'fig_noise_sensitivity.png'));
        fprintf('  Saved: fig_noise_sensitivity.png\n');
    end

    fprintf('Done.\n');
end

File: scripts/plot_results.py
"""Generate publication-quality figures from experiment results."""

import sys, os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'src'))

import json
import numpy as np
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
from matplotlib import rcParams

# IEEE-friendly style
rcParams.update({
    'font.family': 'serif',
    'font.size': 9,
    'axes.labelsize': 10,
    'axes.titlesize': 10,
    'legend.fontsize': 8,
    'xtick.labelsize': 8,
    'ytick.labelsize': 8,
    'figure.dpi': 300,
    'savefig.dpi': 300,
    'savefig.bbox': 'tight',
    'savefig.pad_inches': 0.05,
})

RESULTS_PATH = os.path.join(os.path.dirname(__file__), '..', 'data', 'results', 'all_experiments.json')
FIG_DIR = os.path.join(os.path.dirname(__file__), '..', 'docs', 'paper', 'figures')
os.makedirs(FIG_DIR, exist_ok=True)


def load_results():
    with open(RESULTS_PATH) as f:
        return json.load(f)


def fig_baseline_comparison(data):
    """Bar chart: RMSE comparison across methods."""
    bl = data['baseline']
    methods = ['Odometry', 'EKF\nh=40m', 'EKF\nh=60m', 'EKF\nh=80m', 'EKF\nh=100m', 'Adaptive\nEKF']
    keys = ['odometry', 'ekf_fixed_40', 'ekf_fixed_60', 'ekf_fixed_80', 'ekf_fixed_100', 'adaptive']
    means = [bl[k]['mean_rmse'] for k in keys]
    stds = [bl[k]['std'] for k in keys]

    colors = ['#7f8c8d'] + ['#3498db']*4 + ['#e74c3c']

    fig, ax = plt.subplots(figsize=(3.5, 2.2))
    bars = ax.bar(range(len(methods)), means, yerr=stds, capsize=3,
                  color=colors, edgecolor='black', linewidth=0.5, width=0.65)
    ax.set_xticks(range(len(methods)))
    ax.set_xticklabels(methods, fontsize=7)
    ax.set_ylabel('RMSE (m)')
    ax.set_ylim(0, 8)
    ax.grid(axis='y', alpha=0.3, linewidth=0.5)
    ax.spines['top'].set_visible(False)
    ax.spines['right'].set_visible(False)

    # Value labels
    for bar, val in zip(bars, means):
        ax.text(bar.get_x() + bar.get_width()/2, bar.get_height() + 0.3,
                f'{val:.2f}', ha='center', va='bottom', fontsize=6.5)

    fig.savefig(os.path.join(FIG_DIR, 'baseline_comparison.pdf'))
    fig.savefig(os.path.join(FIG_DIR, 'baseline_comparison.png'))
    plt.close(fig)
    print("  Saved baseline_comparison.pdf/png")


def fig_noise_sensitivity(data):
    """Line plot: RMSE vs pixel noise sigma."""
    nd = data['noise']
    sigmas = sorted([int(k) for k in nd.keys()])
    means = [nd[str(s)]['mean'] for s in sigmas]
    stds = [nd[str(s)]['std'] for s in sigmas]

    fig, ax = plt.subplots(figsize=(3.5, 2.2))
    ax.errorbar(sigmas, means, yerr=stds, fmt='o-', color='#2c3e50',
                markersize=4, capsize=3, linewidth=1.2, markerfacecolor='#3498db',
                markeredgecolor='#2c3e50', markeredgewidth=0.8)
    ax.set_xlabel(r'Pixel noise $\sigma$ (px)')
    ax.set_ylabel('RMSE (m)')
    ax.set_xlim(0, 11)
    ax.set_ylim(0, 0.75)
    ax.grid(alpha=0.3, linewidth=0.5)
    ax.spines['top'].set_visible(False)
    ax.spines['right'].set_visible(False)

    fig.savefig(os.path.join(FIG_DIR, 'noise_sensitivity.pdf'))
    fig.savefig(os.path.join(FIG_DIR, 'noise_sensitivity.png'))
    plt.close(fig)
    print("  Saved noise_sensitivity.pdf/png")


def fig_observability_comparison(data):
    """Side-by-side bar: circular vs straight observability."""
    obs = data['observability']

    fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(3.5, 2.0))

    # Condition number
    trajs = ['Circular', 'Straight']
    conds = [obs['circular']['cond'], obs['straight']['cond']]
    colors = ['#27ae60', '#e74c3c']
    bars1 = ax1.bar(trajs, conds, color=colors, edgecolor='black', linewidth=0.5, width=0.5)
    ax1.set_ylabel('Condition Number')
    ax1.spines['top'].set_visible(False)
    ax1.spines['right'].set_visible(False)
    for bar, val in zip(bars1, conds):
        ax1.text(bar.get_x() + bar.get_width()/2, bar.get_height() + 1,
                 f'{val:.1f}', ha='center', va='bottom', fontsize=7)

    # Mean error
    errors = [obs['circular']['mean_error'], obs['straight']['mean_error']]
    bars2 = ax2.bar(trajs, errors, color=colors, edgecolor='black', linewidth=0.5, width=0.5)
    ax2.set_ylabel('Mean Error (m)')
    ax2.spines['top'].set_visible(False)
    ax2.spines['right'].set_visible(False)
    for bar, val in zip(bars2, errors):
        ax2.text(bar.get_x() + bar.get_width()/2, bar.get_height() + 0.03,
                 f'{val:.2f}', ha='center', va='bottom', fontsize=7)

    fig.tight_layout(w_pad=2.0)
    fig.savefig(os.path.join(FIG_DIR, 'observability_comparison.pdf'))
    fig.savefig(os.path.join(FIG_DIR, 'observability_comparison.png'))
    plt.close(fig)
    print("  Saved observability_comparison.pdf/png")


def fig_eigenvalue_spectrum(data):
    """Bar chart: eigenvalue spectrum for circular vs straight."""
    obs = data['observability']

    fig, ax = plt.subplots(figsize=(3.5, 2.2))
    x = np.arange(3)
    w = 0.3

    eig_c = obs['circular']['eigenvalues']
    eig_s = obs['straight']['eigenvalues']

    ax.bar(x - w/2, eig_c, w, label='Circular', color='#27ae60', edgecolor='black', linewidth=0.5)
    ax.bar(x + w/2, eig_s, w, label='Straight', color='#e74c3c', edgecolor='black', linewidth=0.5)

    ax.set_xticks(x)
    ax.set_xticklabels([r'$\lambda_1$', r'$\lambda_2$', r'$\lambda_3$'])
    ax.set_ylabel('Eigenvalue')
    ax.set_yscale('log')
    ax.legend(frameon=False)
    ax.grid(axis='y', alpha=0.3, linewidth=0.5)
    ax.spines['top'].set_visible(False)
    ax.spines['right'].set_visible(False)

    fig.savefig(os.path.join(FIG_DIR, 'eigenvalue_spectrum.pdf'))
    fig.savefig(os.path.join(FIG_DIR, 'eigenvalue_spectrum.png'))
    plt.close(fig)
    print("  Saved eigenvalue_spectrum.pdf/png")


if __name__ == '__main__':
    print("Generating publication figures...")
    data = load_results()
    fig_baseline_comparison(data)
    fig_noise_sensitivity(data)
    fig_observability_comparison(data)
    fig_eigenvalue_spectrum(data)
    print(f"\nAll figures saved to {FIG_DIR}/")
